training_mod_tui/src/lib.rs

```diff
@@ -1,367 +1,806 @@
+    /// Toggles: Calls previous_list()
+    /// Slider:
+    ///     * Swaps between MinHover and MaxHover
+    ///     * Decrements the selected_min/max if possible
+    pub fn sub_menu_previous_list(&mut self) {
+        match SubMenuType::from_str(self.sub_menu_selected()._type) {
+            SubMenuType::TOGGLE => self.selected_sub_menu_toggles.previous_list(),
+            SubMenuType::SLIDER => match self.selected_sub_menu_slider.state {
+                GaugeState::MinHover => self.selected_sub_menu_slider.state = GaugeState::MaxHover,
+                GaugeState::MaxHover => self.selected_sub_menu_slider.state = GaugeState::MinHover,
+                GaugeState::MinSelected => {
+                    if self.selected_sub_menu_slider.selected_min
+                        > self.selected_sub_menu_slider.abs_min
+                    {
+                        self.selected_sub_menu_slider.selected_min -= 1;
+                    }
+                }
+                GaugeState::MaxSelected => {
+                    if self.selected_sub_menu_slider.selected_max
+                        > self.selected_sub_menu_slider.selected_min
+                    {
+                        self.selected_sub_menu_slider.selected_max -= 1;
+                    }
+                }
+                GaugeState::None => {}
+            },
+        }
+    }
+
+    /// Returns information about the currently selected submenu
+    /// 
+    /// 0: Submenu Title
+    /// 1: Submenu Help Text
+    /// 2: Vec(toggle checked, title) for toggles, Vec(nothing) for slider
+    /// 3: ListState for toggles, ListState::new() for slider
+    pub fn sub_menu_strs_and_states(
+        &mut self,
+    ) -> (&str, &str, Vec<(Vec<(bool, &str)>, ListState)>) {
+        (
+            self.sub_menu_selected().submenu_title,
+            self.sub_menu_selected().help_text,
+            match SubMenuType::from_str(self.sub_menu_selected()._type) {
+                SubMenuType::TOGGLE => self
+                    .selected_sub_menu_toggles
+                    .lists
+                    .iter()
+                    .map(|toggle_list| {
+                        (
+                            toggle_list
+                                .items
+                                .iter()
+                                .map(|toggle| (toggle.checked, toggle.toggle_title))
+                                .collect(),
+                            toggle_list.state.clone(),
+                        )
+                    })
+                    .collect(),
+                SubMenuType::SLIDER => {
+                    vec![(vec![], ListState::default())]
+                }
+            },
+        )
+    }
+
+    /// Returns information about the currently selected slider
+    /// 0: Title
+    /// 1: Help text
+    /// 2: Reference to self.selected_sub_menu_slider
+    pub fn sub_menu_strs_for_slider(&mut self) -> (&str, &str, &DoubleEndedGauge) {
```

Comment:
I'd be down if we have a struct for this output rather than tuple return like `SliderInfo`, similar for toggles. Wonder about your thoughts.

Reply:
Agree that it'd be nice to have in a struct, but I'm going to punt this for later. I'll add a TODO for a cleanup effort.